text/0000-add-ansi-escape-codes.md

```diff
@@ -0,0 +1,63 @@
+- Feature Name: add_ansi_escape_codes
+- Start Date: 2022-02-14
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes the addition of more escape code functions to the ANSI primitive of the term package and the consolidation of the thirty-two existing color functions into four functions.
+
+# Motivation
+
+The purpose of these additional methods is to cover more of the standard ANSI escape codes within the stdlib.
+
+# Detailed design
+
+- Add escape codes for 3-bit, 4-bit, 8-bit, and 24-bit colors for the foreground and background.  
+The naming scheme of the color functions would be `{fg,bg}_{3,4,8,24}bit`.  
```

Comment:
In my quick review of ANSI escape codes, splitting the `{fg,bg}` does not make sense to me. It appears that different `U8` values either set foreground or background. I am using the table for [Colors](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors) in making this comment. 

What is the plan to handle, for example, `fg_4bit(40)` which should be invalid while `bg_4bit(40)` is valid?

Reply:
My reason for wanting to have them split is so that they match the functions for 8-bit and 24-bit colors.  
The plan right now is to return an empty string on invalid values which gives the same effect (of doing nothing) as if those values were printed to the console.

Reply:
I am against such use of default returns -- or "silent failure" -- when it is possible to make the failure loud and thereby avoid errors reaching runtime. I understand the desire to keep the functions consistent though and appreciate that.

Reply:
I just realized that my statement there contradicts another statement I made here where I say I am **not** against default returns. I want to clarify that I am against silent failure when it is possible to make that failure loud, but not against a non-functional default return such as an empty string. If we can make these additions help the user caught mistakes prior to runtime then I want that, if that is going to break existing functionality and an empty return maintains current stability then I am okay with such empty returns. Does that make sense, or have I talked past myself in my attempts to clarify where I stand on this topic?

Reply:
Given the discussion here and the splitting of this RFC into two, I'm thinking that in this one I could leave out the 3-bit and 4-bit color functions and then that leaves the only two functions that could have invalid values passed being the ones for erase in display and erase in line.

The number of valid values for those are small (four for the first and three for the second) and they line up, so making primitives and type unions wouldn't add a ton of types/be a large union like doing one for every color would.

```pony
primitive EraseAfter
primitive EraseBefore
primitive EraseAll
primitive EraseBuffer

type EraseDisplay is (EraseAfter | EraseBefore | EraseAll | EraseBuffer)
type EraseLine is (EraseAfter | EraseBefore | EraseAll)
```

How does that sound?

Reply:
I am not sure I understand. Are you suggesting dropping `{fg,bg}_{3,4}bit` entirely here (which are currently covered by the existing color functions) which would leave only the erase functions with possible invalid inputs? If so, then I can support that. It is a considerable change to make this RFC about adding the 8 and 24 bit colors, but it does free up the design for this initial additions RFC considerably. We could further debate how we want to handle the overlap of breaking changes with existing functionality in the second RFC.

Reply:
Perfect. Then as I said, I support that. I have been feeling a little bad that there are additions (which I have no objection to) tied in with existing overlap (which I have been commenting on) here, but my approval is all or nothing. Thank you for splitting this RFC into separate components. Feel free to take as much or as little of my comments here for the RFC's "Alternative Designs" section.